cozy/codegen/rpython.py

```diff
@@ -0,0 +1,422 @@
+            assert False, "Unhandled binary operation while writing RPython"
+
+    def visit_ECond(self, node):
+        # Quirk: Sometimes cond will be constantly true or false.
+        if isinstance(node, EBool):
+            self.visit(node.then_branch if node.val else node.else_branch)
+
+        with self.parens():
+            self.visit(node.then_branch)
+            self.write(" if ")
+            self.visit(node.cond)
+            self.write(" else ")
+            self.visit(node.else_branch)
+
+    def visit_EMap(self, node):
+        f = node.f
+        # Quirk: Sometimes we will see (map (\x -> x) xs) and we want to just
+        # have xs instead.
+        if f.arg == f.body:
+            self.visit(node.e)
+            return
+
+        self.write("[")
+        self.visit(f.body)
+        self.write(" for ")
+        # Hack: Write the lambda's arg as if it were an expr. This is close
+        # enough for now but definitely wrong. ~ C.
+        self.visit(f.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write("]")
+
+    def visit_EFilter(self, node):
+        p = node.p
+        self.write("[")
+        # Double hack! Abuse the arg again.
+        self.visit(p.arg)
+        self.write(" for ")
+        self.visit(p.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write(" if ")
+        self.visit(p.body)
+        self.write("]")
+
+    def visit_EArgMax(self, node):
+        f = node.f
+        self.needs["argMax"] = True
+        self.write("_argMax([")
+        self.visit(f.body)
+        self.write(" for ")
+        self.visit(f.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write("])")
+
+    def visit_EArgMin(self, node):
+        f = node.f
+        self.needs["argMin"] = True
+        self.write("_argMin([")
+        self.visit(f.body)
+        self.write(" for ")
+        self.visit(f.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write("])")
+
+    def visit_EMakeMap2(self, node):
+        self.needs["makeMap"] = True
+        f = node.value
+        self.write("_makeMap([(")
+        self.visit(f.arg)
+        self.write(", ")
+        self.visit(f.body)
+        self.write(") for ")
+        self.visit(f.arg)
+        self.write(" in ")
+        self.visit(node.e)
+        self.write("])")
+
+    def visit_EMapGet(self, node):
+        self.visit(node.map)
+        self.write("[")
+        self.visit(node.key)
+        self.write("]")
+
+    def visit_EHasKey(self, node):
+        with self.parens():
+            self.visit(node.key)
+            self.write(" in ")
+            self.visit(node.map)
+
+    def visit_ETuple(self, node):
+        with self.parens():
+            for e in node.es:
+                self.visit(e)
+                self.write(", ")
+
+    def visit_ETupleGet(self, node):
+        # Quirk: Accesses to x.0 are phrased as x.val.0 for some reason.
+        if node.n == "val":
```

Comment:
This should be dead code. If you ever saw this happen it is a bug.

Reply:
I may have cargo-culted this from the record-handling code. On reflection, I think tuples can be RPython tuples, without reservation, but records... I'm on the fence, but I think the right option is to generate an "inner" class (at top level, because RPython reasons) for each seen record type.